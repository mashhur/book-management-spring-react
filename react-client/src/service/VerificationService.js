const axios = require("axios");

const BASE_URL = "http://localhost:8080/api/verify";

class VerficationService {
	getOtp(phoneNumber) {
		return axios.post(BASE_URL + "/get-otp", {
			phoneNumber: phoneNumber,
		});
	}
	verifyOtp(body) {
		return axios.post(BASE_URL + "/verify-otp", body);
	}
}

module.exports = new VerficationService();
