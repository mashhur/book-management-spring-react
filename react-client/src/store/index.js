import { configureStore } from "@reduxjs/toolkit";
import locationReducer from "./slices/locationSlice";
import userReducer from "./slices/userSlice";

export const store = configureStore({
	reducer: {
		location: locationReducer,
		user: userReducer,
	},
});
