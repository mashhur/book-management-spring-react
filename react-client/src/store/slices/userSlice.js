import { createSlice } from "@reduxjs/toolkit";

const inituser = JSON.parse(localStorage.getItem("user"));
const user = inituser || {
	cid: "",
	fullName: "",
	phoneNumber: "",
};
const initialState = {
	auth: localStorage.getItem("auth") || false,
	user: user,
	address: {
		aid: "",
		customerName: "",
		phoneNumber: "",
		address: "",
		mstate: "",
		pincode: "",
		customerId: "",
	},
	books: [],
	cart: [],
};

export const userSlice = createSlice({
	name: "user",
	initialState,
	reducers: {
		changeAuth: (state) => {
			state.auth = true;
		},
		setUser: (state, action) => {
			state.user = action.payload;
		},
		setAddress(state, action) {
			state.address = action.payload;
		},
		addBook(state, action) {
			state.books = [];
			state.books.push(action.payload);
		},
		addBooktoCart(state, action) {
			state.cart.push(action.payload);
		},
		addtoBuy(state) {
			const cart = state.cart;
			state.books = cart;
			state.cart = [];
		},
		clearOrderDetails(state) {
			state.books = [];
			state.address = {};
		},
		removeBook(state, action) {
			state.cart.splice(action.payload, 1);
		},
		clearAuth(state) {
			state.auth = false;
			state.user = {
				cid: "",
				fullName: "",
				phoneNumber: "",
			};
		},
	},
});

export const {
	changeAuth,
	setUser,
	setAddress,
	addBook,
	clearOrderDetails,
	removeBook,
	addBooktoCart,
	addtoBuy,
	clearAuth,
} = userSlice.actions;

export default userSlice.reducer;
