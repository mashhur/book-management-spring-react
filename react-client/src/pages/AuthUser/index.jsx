import React from "react";
import { useDispatch } from "react-redux";
import { useSelector } from "react-redux";
import CustomerService from "../../service/CustomerService";
import VerificationService from "../../service/VerificationService";
import {
	changeAuth,
	setUser,
} from "../../store/slices/userSlice";

function AuthUser(props) {
	const location = useSelector((state) => state.location);
	const dispatch = useDispatch();

	const [authenticationState, setauthenticationState] =
		React.useState({
			stage: "input",
			phoneNumber: null,
			buttonText: "Send an OTP",
			buttonDisabled: false,
			otp: null,
			error: null,
			hash: null,
		});

	const onChange = (e) => {
		setauthenticationState({
			...authenticationState,
			[e.target.name]: e.target.value,
			error: null,
		});
	};

	const {
		stage,
		phoneNumber,
		buttonText,
		buttonDisabled,
		otp,
		error,
	} = authenticationState;

	const clickHandler = () => {
		if (stage === "input") {
			//verfiy number

			if (phoneNumber.length !== 10) {
				setauthenticationState({
					...authenticationState,
					error: "Please enter a valid phone number",
				});
				return;
			}

			VerificationService.getOtp(phoneNumber).then(
				(res) => {
					console.log(res.data);
					if (res.status === 200) {
						setauthenticationState({
							...authenticationState,
							stage: "verification",
							buttonText: "Verify OTP",
							buttonDisabled: false,
							error: null,
							hash: res.data,
						});
					} else {
						setauthenticationState({
							...authenticationState,
							error: "Something went wrong",
							buttonDisabled: false,
							buttonText: "Send an OTP",
						});
					}
				}
			);
		} else {
			if (otp.length !== 4) {
				setauthenticationState({
					...authenticationState,
					error: "Please enter a valid 4 digit code",
				});
				return;
			}

			if (authenticationState.hash === null) {
				setauthenticationState({
					...authenticationState,
					error: "Something went wrong",
					buttonDisabled: false,
					buttonText: "Send an OTP",
				});
				return;
			}

			const verifyBody = {
				hash: authenticationState.hash,
				otp: otp,
				phoneNumber: phoneNumber,
			};

			setauthenticationState({
				...authenticationState,
				buttonText: "Please wait...",
				buttonDisabled: true,
				error: null,
			});

			VerificationService.verifyOtp(verifyBody)
				.then((res) => {
					if (res.status === 200) {
						CustomerService.getCustomer(phoneNumber).then(
							(res) => {
								console.log(res);
								if (res.data) {
									localStorage.setItem(
										"user",
										JSON.stringify(res.data)
									);
									localStorage.setItem("auth", true);
									dispatch(setUser(res.data));
									dispatch(changeAuth(true));
									setTimeout(() => {
										// navigate(location.location.pathname);
									}, 1000);
								} else {
									// get name
									let name = prompt("Enter your name");
									if (name) {
										const user = {
											fullName: name,
											phoneNumber: phoneNumber,
										};
										CustomerService.addCustomer(user).then(
											(res) => {
												localStorage.setItem(
													"user",
													JSON.stringify(res.data)
												);
												localStorage.setItem("auth", true);
												dispatch(setUser(res.data));
												dispatch(changeAuth(true));
												setTimeout(() => {
													// navigate();
													// location.location.pathname
												}, 1000);
											}
										);
									} else {
										setauthenticationState({
											...authenticationState,
											error: "Please enter a valid name",
										});
									}
								}
							}
						);
					} else {
						setauthenticationState({
							...authenticationState,
							error: "Something went wrong",
							buttonDisabled: false,
							buttonText: "Try again.",
						});
					}
				})
				.catch((e) => {
					console.log(e);
					setauthenticationState({
						...authenticationState,
						error: "Invalid otp, try again.",
						buttonDisabled: false,
						buttonText: "send Again.",
					});
				});
		}
	};

	return (
		<div
			data-test="component-auth"
			style={{ height: "100vh" }}
			className="w-100 
		 d-flex justify-content-center">
			<div
				style={{ borderRadius: "1.5rem", width: "35%" }}
				className="card my-auto shadow-lg border-0  ">
				<div className="card-body mt-4 py-4 px-5 ">
					{stage === "input" && (
						<div data-test="component-phone-number">
							<h3 className="mb-0">Hey, sign in</h3>
							<p className="text-muted mt-0 me-3">
								to proceed ahead...
							</p>
							<div className="mb-3">
								<label
									htmlFor="exampleInputPassword1"
									className="form-label">
									Please enter you phone number
								</label>
								<input
									style={{
										borderRadius: ".5rem",
										padding: ".63rem",
									}}
									name="phoneNumber"
									type="text"
									maxLength="10"
									pattern="[0-9]{10}"
									className={`form-control ${
										error ? "is-invalid" : ""
									}`}
									id="exampleInputPassword1"
									onChange={onChange}
								/>
								{error && (
									<div
										data-test="error-comp"
										className="form-text text-danger"
										role="alert">
										{error}
									</div>
								)}
								{!error && (
									<div id="emailHelp" className="form-text">
										We'll send an otp on this number for
										verification.
									</div>
								)}
							</div>
						</div>
					)}

					{stage === "verification" && (
						<>
							<h3 className="mb-0">Verification.</h3>
							<p className="text-muted mt-0 me-3">
								Please provide the otp we've sent you!
							</p>
							<div className="mb-3 pt-2">
								<label className="form-label">
									Enter the otp
								</label>{" "}
								<input
									style={{
										borderRadius: ".5rem",
										padding: ".63rem",
									}}
									name="otp"
									type="text"
									placeholder="0-0-0-0"
									maxLength="4"
									pattern="[0-9]{4}"
									className={`form-control text-center fs-1 text-muted ${
										error ? "is-invalid" : ""
									}`}
									onChange={onChange}
								/>
								{error && (
									<div
										data-test="error-comp"
										className="form-text text-danger"
										role="alert">
										{error}
									</div>
								)}
							</div>
						</>
					)}

					<button
						onClick={clickHandler}
						disabled={buttonDisabled}
						style={{
							borderRadius: ".5rem",
							padding: ".63rem",
						}}
						type="submit"
						className="btn w-100 mt-2 mb-4 
					btn-primary">
						{buttonText}
					</button>
				</div>
			</div>
		</div>
	);
}

export default AuthUser;
