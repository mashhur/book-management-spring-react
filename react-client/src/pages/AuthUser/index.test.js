import React from "react";
import { mount, shallow } from "enzyme";
import AuthUser from "./";
import moxios from "moxios";
import { findByTestAttr } from "../../tests/testUtils";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { createTestStore } from "../../store";
import VerificationService from "../../service/VerificationService";
import { setLocation } from "../../store/slices/locationSlice";

const defaultState = {
	stage: "input",
	phoneNumber: null,
	buttonText: "Send an OTP",
	buttonDisabled: false,
	otp: null,
	error: null,
	hash: null,
};

const store = createTestStore();
const setup = () => {
	return mount(
		<Provider store={store}>
			<AuthUser />
		</Provider>
	);
};

describe("testing the authentication", () => {
	beforeAll(() => {
		moxios.install();
		store.dispatch(setLocation({ pathname: "/" }));
	});
	afterAll(() => {
		moxios.uninstall();
	});

	it("renders without any error", async () => {
		const wrapper = setup();
		const component = await findByTestAttr(
			wrapper,
			"component-auth"
		);
		expect(component.length).toBe(1);
	});
	describe("when `getOtp()` is called", () => {
		it("return the correct repsonse", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				req.respondWith({
					status: 200,
					response: "thisisthehash",
				});
			});
			return VerificationService.getOtp("123456").then(
				(respsonse) => {
					expect(respsonse.data).toEqual("thisisthehash");
				}
			);
		});

		//response error has been tested.
		it("return error response", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				const errResp = {
					status: 500,
					response: {
						msg: "The server is unable to respond",
					},
				};
				req.reject(errResp);
			});

			return VerificationService.getOtp("123456").catch(
				(error) => {
					// eslint-disable-next-line jest/no-conditional-expect
					expect(error.status).toEqual(500);
				}
			);
		});
	});

	it("verifying the inputs should change state", () => {
		const functionCaller = jest.fn();
		React.useState = jest.fn(() => [
			defaultState,
			functionCaller,
		]);
		const wrapper = setup();
		const input = wrapper.find("input");
		input.simulate("change");
		expect(functionCaller).toHaveBeenCalled();
	});

	describe("when send otp button is clicked", () => {
		beforeEach(() => {
			jest.spyOn(VerificationService, "getOtp");
		});

		it("if phone number available call `getOtp`", async () => {
			const state = {
				...defaultState,
				phoneNumber: "1234567890",
			};
			React.useState = jest.fn(() => [state, jest.fn()]);
			const wrapper = setup();
			const button = wrapper.find("button");
			button.simulate("click");
			expect(VerificationService.getOtp).toHaveBeenCalled();
		});
		it("if phone number incorrect don't call `getOtp`", async () => {
			const state = {
				...defaultState,
				phoneNumber: "123456789",
			};
			React.useState = jest.fn(() => [state, jest.fn()]);
			const wrapper = setup();
			const button = wrapper.find("button");
			button.simulate("click");
			expect(
				VerificationService.getOtp
			).not.toHaveBeenCalled();
		});
	});

	describe("when verify otp button is clicked", () => {
		beforeEach(() => {
			jest.spyOn(VerificationService, "verifyOtp");
		});

		afterEach(() => {
			jest.clearAllMocks();
		});
		describe("if otp & hash available call `verifyOtp`", () => {
			it("should request the api", () => {
				const state = {
					...defaultState,
					stage: "verification",
					otp: "1234",
					phoneNumber: "1234567890",
					hash: "testhash",
				};
				React.useState = jest.fn(() => [state, jest.fn()]);
				const wrapper = setup();

				const button = wrapper.find("button");
				button.simulate("click");
				expect(
					VerificationService.verifyOtp
				).toHaveBeenCalled();
			});
		});

		it("if otp not available dont call `verifyOtp`", async () => {
			const state = {
				...defaultState,
				stage: "verification",
				otp: "",
				phoneNumber: "1234567890",
				hash: null,
			};
			React.useState = jest.fn(() => [state, jest.fn()]);
			const wrapper = setup();

			const button = wrapper.find("button");
			button.simulate("click");
			expect(
				VerificationService.verifyOtp
			).not.toHaveBeenCalled();
		});
	});

	describe("rendering ui components", () => {
		it("should render phone number comp by default", async () => {
			React.useState = jest.fn(() => [
				defaultState,
				jest.fn(),
			]);
			const wrapper = setup();
			const component = await findByTestAttr(
				wrapper,
				"component-phone-number"
			);
			expect(component.length).toBe(1);
		});

		it("should render the correct snapshot by default", () => {
			React.useState = jest.fn(() => [
				defaultState,
				jest.fn(),
			]);
			const wrapper = setup();
			expect(wrapper).toMatchSnapshot();
		});

		it("should render otp comp", async () => {
			React.useState = jest.fn(() => [
				{
					...defaultState,
					stage: "verification",
				},
				jest.fn(),
			]);
			const wrapper = setup();
			const component = await findByTestAttr(
				wrapper,
				"component-phone-number"
			);
			expect(component.length).toBe(0);
		});

		it("should render the correct snapshot of otp", () => {
			React.useState = jest.fn(() => [
				{
					...defaultState,
					stage: "verification",
				},
				jest.fn(),
			]);
			const wrapper = setup();
			expect(wrapper).toMatchSnapshot();
		});
	});
});
