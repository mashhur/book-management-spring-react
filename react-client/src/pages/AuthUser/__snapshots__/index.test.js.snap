// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the authentication rendering components should render the correct snapshot 1`] = `ReactWrapper {}`;

exports[`testing the authentication rendering components should render the correct snapshot by default 1`] = `ReactWrapper {}`;

exports[`testing the authentication rendering components should render the correct snapshot of otp 1`] = `ReactWrapper {}`;

exports[`testing the authentication rendering ui components should render the correct snapshot by default 1`] = `ReactWrapper {}`;

exports[`testing the authentication rendering ui components should render the correct snapshot of otp 1`] = `ReactWrapper {}`;
